### ISSUE TITLE
Ubuntu 16.04 / Gpu 1080 Version compile Error (Tensorflow installation)

### ISSUE THREAD

**Issue by sungjin712** [tensorflow/tensorflow#3760]
Hello, I recently got a new machine gpu 1080 and I did follow most of instructions via http://textminingonline.com/dive-into-tensorflow-part-iii-gtx-1080-ubuntu16-04-cuda8-0-cudnn5-0-tensorflow.
Anyone successfully installed Ubuntu 16.04 GPU with that version?? Let me know

But the Issue is it doesn't let me complie via bazel.

The following code...

ERROR: /home/ryan/git_ryan/tensorflow/tensorflow/core/kernels/BUILD:1655:1: undeclared inclusion(s) in rule '//tensorflow/core/kernels:dense_update_ops_gpu':
this rule is missing dependency declarations for the following files included by 'tensorflow/core/kernels/dense_update_ops_gpu.cu.cc':
  '/usr/local/cuda-8.0/include/cuda_runtime.h'
  '/usr/local/cuda-8.0/include/host_config.h'
  '/usr/local/cuda-8.0/include/builtin_types.h'
  '/usr/local/cuda-8.0/include/device_types.h'
  '/usr/local/cuda-8.0/include/host_defines.h'
  '/usr/local/cuda-8.0/include/driver_types.h'
  '/usr/local/cuda-8.0/include/surface_types.h'
  '/usr/local/cuda-8.0/include/texture_types.h'
  '/usr/local/cuda-8.0/include/vector_types.h'
  '/usr/local/cuda-8.0/include/library_types.h'
  '/usr/local/cuda-8.0/include/channel_descriptor.h'
  '/usr/local/cuda-8.0/include/cuda_runtime_api.h'
  '/usr/local/cuda-8.0/include/cuda_device_runtime_api.h'
  '/usr/local/cuda-8.0/include/driver_functions.h'
  '/usr/local/cuda-8.0/include/vector_functions.h'
  '/usr/local/cuda-8.0/include/vector_functions.hpp'
  '/usr/local/cuda-8.0/include/common_functions.h'
  '/usr/local/cuda-8.0/include/math_functions.h'
  '/usr/local/cuda-8.0/include/math_functions.hpp'
  '/usr/local/cuda-8.0/include/math_functions_dbl_ptx3.h'
  '/usr/local/cuda-8.0/include/math_functions_dbl_ptx3.hpp'
  '/usr/local/cuda-8.0/include/cuda_surface_types.h'
  '/usr/local/cuda-8.0/include/cuda_texture_types.h'
  '/usr/local/cuda-8.0/include/device_functions.h'
  '/usr/local/cuda-8.0/include/device_functions.hpp'
  '/usr/local/cuda-8.0/include/device_atomic_functions.h'
  '/usr/local/cuda-8.0/include/device_atomic_functions.hpp'
  '/usr/local/cuda-8.0/include/device_double_functions.h'
  '/usr/local/cuda-8.0/include/device_double_functions.hpp'
  '/usr/local/cuda-8.0/include/sm_20_atomic_functions.h'
  '/usr/local/cuda-8.0/include/sm_20_atomic_functions.hpp'
  '/usr/local/cuda-8.0/include/sm_32_atomic_functions.h'
  '/usr/local/cuda-8.0/include/sm_32_atomic_functions.hpp'
  '/usr/local/cuda-8.0/include/sm_35_atomic_functions.h'
  '/usr/local/cuda-8.0/include/sm_60_atomic_functions.h'
  '/usr/local/cuda-8.0/include/sm_60_atomic_functions.hpp'
  '/usr/local/cuda-8.0/include/sm_20_intrinsics.h'
  '/usr/local/cuda-8.0/include/sm_20_intrinsics.hpp'
  '/usr/local/cuda-8.0/include/sm_30_intrinsics.h'
  '/usr/local/cuda-8.0/include/sm_30_intrinsics.hpp'
  '/usr/local/cuda-8.0/include/sm_32_intrinsics.h'
  '/usr/local/cuda-8.0/include/sm_32_intrinsics.hpp'
  '/usr/local/cuda-8.0/include/sm_35_intrinsics.h'
  '/usr/local/cuda-8.0/include/surface_functions.h'
  '/usr/local/cuda-8.0/include/texture_fetch_functions.h'
  '/usr/local/cuda-8.0/include/texture_indirect_functions.h'
  '/usr/local/cuda-8.0/include/surface_indirect_functions.h'
  '/usr/local/cuda-8.0/include/device_launch_parameters.h'
  '/usr/local/cuda-8.0/include/cuda_fp16.h'
  '/usr/local/cuda-8.0/include/math_constants.h'
  '/usr/local/cuda-8.0/include/curand_kernel.h'
  '/usr/local/cuda-8.0/include/curand.h'
  '/usr/local/cuda-8.0/include/curand_discrete.h'
  '/usr/local/cuda-8.0/include/curand_precalc.h'
  '/usr/local/cuda-8.0/include/curand_mrg32k3a.h'
  '/usr/local/cuda-8.0/include/curand_mtgp32_kernel.h'
  '/usr/local/cuda-8.0/include/cuda.h'
  '/usr/local/cuda-8.0/include/curand_mtgp32.h'
  '/usr/local/cuda-8.0/include/curand_philox4x32_x.h'
  '/usr/local/cuda-8.0/include/curand_globals.h'
  '/usr/local/cuda-8.0/include/curand_uniform.h'
  '/usr/local/cuda-8.0/include/curand_normal.h'
  '/usr/local/cuda-8.0/include/curand_normal_static.h'
  '/usr/local/cuda-8.0/include/curand_lognormal.h'
  '/usr/local/cuda-8.0/include/curand_poisson.h'
  '/usr/local/cuda-8.0/include/curand_discrete2.h'.
nvcc warning : option '--relaxed-constexpr' has been deprecated and replaced by option '--expt-relaxed-constexpr'.
nvcc warning : option '--relaxed-constexpr' has been deprecated and replaced by option '--expt-relaxed-constexpr'.
Target //tensorflow/cc:tutorials_example_trainer failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 50.107s, Critical Path: 49.88s

For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

### Environment info
Operating System: Linux 64bit / GPU 1080

Installed version of CUDA and cuDNN:  Cuda 8 / cuDNN: 8.0
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):

**Comment by Mazecreator** [comment]
You need to update the CROSSTOOLS file to see this CUDA includes:
tensorflow/third_party/gpus/crosstool/CROSSTOOL

Around line 65, add:
  cxx_builtin_include_directory: "/usr/local/cuda-8.0/include"

**Comment by sungjin712** [comment]
@Mazecreator  Thank you for your answer. But I still have pywrap error...which doesnt work for my pc.

**Comment by falcondai** [comment]
+1

**Comment by Mazecreator** [comment]
I am not certain I know what pywrap error you are having, but I have Ubuntu 16.04 with CUDA 8.0 / cuDNN 5 running on my  system with no problems.

I did manually install the latest NVIDIA 367.35.0 Video Driver (.run) and the (.run) file for CUDA 8.0 as well.  Make sure you do not ignore the gcc version error.  I did that and everything compiled properly, but python failed to execute a tensor graph.  I had to go back and create a symbolic link from gcc-4.9 to gcc for the CUDA build.

A few things to try if you haven't already as I am guessing a bit since I don't have the pywrap details:
- make sure you do a "bazel clean" ( you might need to use the --expunge_async option as well)
- if you are running python from the "tensorflow" repository directory, it is known not to work, change to a different director (not sure why, but it is true)
- Search for other CUDA 8.0 issues under tensorflow as there are a few and they may help

If that doesn't help, I would need more details for your current error.

**Comment by girving** [comment]
@sungjin712 If you want help with a pywrap error, you'll have to show us the error.

**Comment by falcondai** [comment]
solved my own problem by using the official docker image to compile a wheel against cudnn 5 and then copying out to the host. i think adding a pointer to docker in the `building from source` section in doc will be super helpful.

**Comment by sungjin712** [comment]
@Mazecreator 
I also followed all the instructions from (http://textminingonline.com/dive-into-tensorflow-part-iii-gtx-1080-ubuntu16-04-cuda8-0-cudnn5-0-tensorflow) which I assumed very similar to what you explained. There should be some compile errors and let me try again and update the error if it doesn't work.  Good to heard that there are some people actually did the installation.  Thanks for your help anyway!

**Comment by sungjin712** [comment]
@girving  Yes, I will try the full procedure again and upload the error if it doesn't work.

**Comment by sungjin712** [comment]
@falcondai Where can I get those sources? I guess I need to learn how to use docker well too.

**Comment by falcondai** [comment]
@sungjin712 look at the CI build folder here: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/ci_build
(i do not recommend trying the build without docker which defeats the whole purpose)

The main idea is that we run a standard CI (Continuous Integration) build in a docker image (clean, standard environment) and then copy out the pip wheel file:
- install docker and nvidia-docker. have the docker daemon running
- make sure you have checkout the intended tensorflow version in your cloned git repo, e.g. `git checkout v0.10.0rc0`.
- go to the ci_build folder and copy this file from my gist: https://gist.github.com/falcondai/7e159cc3cfa1529996f567ca284b00a7#file-docker_build-sh (the original build script removes the container after each build)
- make sure [the first line in Dockerifle.gpu](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/ci_build/Dockerfile.gpu#L1) is the cuda + cudnn version you want: for cuda 8 and cudnn 5, it should be `FROM nvidia/cuda:8.0-cudnn5-devel`. (Note that this image uses ubuntu 14.04 as base, but i think the wheel file will likely work for ubuntu 16.04. if you really want to *build* in ubuntu 16.04, you need to look into first building your own cuda image, which should be a matter of changing a line in the cuda image Dockerfile and then building, since all public nvidia/cuda images are built on 14.04: https://hub.docker.com/r/nvidia/cuda/)
- `$ ./docker_build.sh gpu tensorflow/tools/ci_build/builds/pip.sh gpu`. (this will take a while)
- look up the container id of your `tf_ci.gpu` by `docker ps`
- copy out the built wheel file `docker cp <container_id>:pip_test/whl/tensorflow*.whl <host-path>`
- pip install this wheel file `$ pip install -U <host-path>`

**Comment by sungjin712** [comment]
I tried again and It seems like after compiling the file via bazel, 
"bazel-bin/tensorflow/cc/tutorials_example_trainer --use_gpu" command returns error like below. It seems like errors from my compiling issues.

F tensorflow/cc/tutorials/example_trainer.cc:80] Check failed: ::tensorflow::Status::OK() == (root.ToGraphDef(&def)) (OK vs. Unimplemented: Explict cast of a non-empty tensor not implemented yet)
F tensorflow/cc/tutorials/example_trainer.cc:80] Check failed: ::tensorflow::Status::OK() == (root.ToGraphDef(&def)) (OK vs. Unimplemented: Explict cast of a non-empty tensor not implemented yet)
F tensorflow/cc/tutorials/example_trainer.cc:80] Check failed: ::tensorflow::Status::OK() == (root.ToGraphDef(&def)) (OK vs. Unimplemented: Explict cast of a non-empty tensor not implemented yet)
Aborted (core dumped)

------

INFO: From Compiling tensorflow/core/kernels/cwise_op_gpu_atan.cu.cc:
nvcc warning : option '--relaxed-constexpr' has been deprecated and replaced by option '--expt-relaxed-constexpr'.
nvcc warning : option '--relaxed-constexpr' has been deprecated and replaced by option '--expt-relaxed-constexpr'.

Always thanks for your support

**Comment by sungjin712** [comment]
@falcondai Thanks for alternative solutions. It takes little bit of time for me to understand following those instructions since I didn't use docker, but I will try that!

**Comment by girving** [comment]
@sungjin712 That "Explicit cast" issue was recently fixed; see https://github.com/tensorflow/tensorflow/issues/3752.  Are you at HEAD?

**Comment by sungjin712** [comment]
@girving Yeah, I saw that comment before and just ignore and install tensorflow via pip, but it has pywrap error which dose not work for me. By the way, what do you mean by "at HEAD"???

**Comment by girving** [comment]
@sungjin712 By "at HEAD" I mean from the current source, as opposed to from a release such as 0.10.

**Comment by sungjin712** [comment]
@girving I have tensorflow-0.10.0rc0-py2-none-any.whl (seems like 10.0rc) version complied from current source. So, do I need to change the version or sth? It seems like recent ver. since I cloned the tensorflow repository last week.

**Comment by girving** [comment]
@sungjin712 #3752 was fixed on Friday, so last week isn't recent enough. :)

**Comment by sungjin712** [comment]
@girving Lol, thanks for your clarification. Let me re-download and try again!

**Comment by sungjin712** [comment]
@girving Thank you so much, It finally works. Everything I did was right except I needed to download current version of tensorflow source.
